### DISCUSSION TITLE
About `ReadableStream` as helper in `SSE` (`httpSubscriptionLink`) (& some other questions re. subscriptions)

### DISCUSSION THREAD

**Discussion by Katli95** [trpc/trpc#5897] [Q&A] [answered]
I'm implementing my first subscription, intending to use `sse`.

I was reading through the `next-sse-chat` example and ran into the following code in [`server/routers/post.ts:105`](https://github.com/trpc/trpc/blob/next/examples/next-sse-chat/src/server/routers/post.ts#L105)
```ts
 // We use a readable stream here to prevent the client from missing events
const stream = new ReadableStream<PostType>({
    async start(controller) {
        const onAdd = (channelId: string, data: PostType) => {
            if (channelId === opts.input.channelId) {
              controller.enqueue(data);
            }
        };
        ee.on('add', onAdd);
        //...
});

for await (const post of streamToAsyncIterable(stream)) {
    yield sse({
        // ...    
    });
}
```

I've got a couple of questions:
1. What exactly is the benefit of using the stream as intermediary? Is the client ever expected to "miss events" without it? I.e. under what conditions/requirements should I be using that in lieu of just a basic `yield`?
2. How does one make sure the `.subscription()` stops, i.e. what methods exist for cancelling/unsubscribing? (related to #5840)
3. What's the error flow like within the body of the route? I saw #5840 had some follow up questions relating to that (albeit they were using an `observable` pattern, which I'm not familiar with and don't see mentioned in the docs)

**Comment by KATT** [comment]
1. When it's called with `lastEventId` the `getPostsSince(x)` can have a race condition with starting listening to the subscription
2. We listen to requests being aborted and call `return()` on the iterator
3. Unsure what you mean and I'm on my phone so unable to click through much atm

**Reply by Katli95** [reply]
Thanks for answering so quick, and on your phone! I'm sorry, but you could explain in more detail, I'd be happy to then move the answers into a PR amending the relative docs.

1. I don't see where the race condition could arise, could you explain in more detail?
    1. _Edit: Do you mean because `getPostsSince(x)` can take long enough for the connection request to timeout before the code `yields`? If so, I still don't understand how the `Stream` is helping :/_
3. Is this dependent on the front-end invoking `.unsubscribe()` or flipping `skipToken` explicitly, also is it enough to un-mount the component calling `.useSubscription()`? I'm trying to understand if there's any risk of _memory leaks_ or the route not halting if the client doesn't explicitly close (e.g. on crash/navigate away)
4. Basically if throwing in the function propagates the error to the `onError` or if something specific is needed (or possible for that matter), #5840 had problems using `observable => sub.error()` not showing up on the frontend.

**Reply by KATT** [reply, marked as answer]
1. 

```ts


async function* (opts) {
  const newItemsSinceCursor = await db.query.Post.findMany({
    where: (fields, ops) =>
      ops.and(
        ops.eq(fields.channelId, opts.input.channelId),
        lastMessageCursor
          ? ops.gt(fields.createdAt, lastMessageCursor)
          : undefined,
      ),
    orderBy: (fields, ops) => ops.asc(fields.createdAt),
  });

  for (const item of newItemsSinceCursor) {
    yield see({
      id: item.id,
      data: item,
    })
  }
  // ❌ even if it's rare/unlikely events might have been emitted while the above code was running
  
  for await (const [post] of ee.toIterable('add')) {
    yield sse({
      id: post.id,
      data: post,
    })
  }
}
```

There might be a nicer way of doing this that doesn't require a ReadableStream - feel free to do PRs. Also, feel free to add more comments/etc that would be helpful for yourself

2. It should happen on any "disconnect event" like the client aborting the request. I'd advise you to play with it. If it doesn't clean up, it's a bug.

3. Throwing in the function should propagate to `onError()` in the backend, but the event is **not** serialized and sent to the frontend right now

**Reply by Katli95** [reply]
Aaaa, right, awesome! Thank you so much for the quick, detailed reply. Loving the library btw, I attribute 85% of my projects stability to `trpc` 🔥
